### DISCUSSION TITLE
ADC accuracy

### DISCUSSION THREAD

**Discussion by dvarrel** [micropython/micropython#11013] [ESP32]
Hi,
I tried to work with adc value, but i can't reach accurate measurements.
what's wrong with my code ?

```
  import machine, esp32, time, os
  from machine import Pin, Timer, ADC, DAC
  
  led2 = Pin(2, Pin.OUT)    # create output pin on GPIO2
  adcPin = 36
  dacPin = 25
  adc = ADC(Pin(adcPin))        # create an ADC object acting on a pin
  adc.atten(ADC.ATTN_11DB) #max adc 2.6V
  dac = DAC(Pin(dacPin))
  Udac_max = 3.17 #max DAC voltage mesured
  N = int(1.0/Udac_max*255)
  dac.write(N)
  
  
  def timer(led):
      if led.value():
          led.value(0)# set pin to "off" (low) level
      else:
          led.value(1) # set pin to "on" (high) level
      print("read_uv():{:.2f}V".format(adc.read_uv()/1e6))
      N = adc.read_u16()
      print("readu16():{} U:{:.2f}V".format(N,(1.1/(2**16-1))*N))
      N = adc.read()
      print("read():{} U:{:.2f}V".format(N, (3.3/(2**12-1))*N))
      
          
  def start():
      print(os.uname())
      print(machine.freq()//10**6,"Mhz")
      print("hall sensor:",esp32.hall_sensor())     # read the internal hall sensor
      print("cpu:{:.1f}°C".format((esp32.raw_temperature()-32)*5/9)) # read the internal temperature of the MCU, in Fahrenheit
  
      t0 = Timer(0)
      t0.init(period=2000, mode=Timer.PERIODIC, callback=lambda x:timer(led2))
      
  if __name__=="__main__":
      start()


```
out:
```
read_uv():3.07V
readu16():61438 U:1.03V
read():3837 U:3.09V
```

**Comment by robert-hh** [comment]
A few comments:
- at an ATTEN value of 11 dB, the range is 0 to ~3.3V. The documentation is wrong, as can easy be calculated by 10**(ATTEN/20).
- In the printout for read_u16, you use 1.1V as reference instead of ~3.3. If you multiply the 1.03 by 3, you'll be at 3.09 then.

So I think the 3.07 or 3.09 readings are consistent. The DAC value you set has to be checked with a voltmeter.

**Reply by robert-hh** [reply]
Checking back at the espressif docs and my selection of boards, there is an interesting difference between boards:
- the old ESP32 have an initial range of 0-1.1V, and 0-3.3V at 11db attenuation, and a 12 bit resolution
- a ESP32S2 board has an base range of 0-0,85V, and 0-2.5V at 11db attenuation
- a ESP32C3 has a base range of 0-0,88 V

Interestingly, the espressif documentation makes no difference between mcu types, but an Arduino doc does, but even that one does not mention the older ESP32 models.  https://docs.espressif.com/projects/arduino-esp32/en/latest/api/adc.html

**Comment by dvarrel** [comment]
Thank you, yes i measure 1.02v with my voltmeter at dac out

Le 12 mars 2023 08:23:38 GMT+01:00, Robert Hammelrath ***@***.***> a écrit :

**Comment by dvarrel** [comment]
Thé micropython doc said read_UV() returns directly voltage with calibration use of fuse

Le 12 mars 2023 08:47:13 GMT+01:00, Robert Hammelrath ***@***.***> a écrit :

**Reply by robert-hh** [reply]
It returns the value before attenuation as microvolt referred to the reference voltag. Connecting to a sufficiently high input voltag, it shoud return the ref voltage. For my esp32S2 it is 838000 or 0.838Vref, for the old ESP32 it is 1039000 or 1.039Vref. So it seems that with your code the adc.read_uv() reading is strange.

**Reply by robert-hh** [reply]
With a slightly modified code I read on my old ESP32:
```
read_uv():1.00V
readu16():17156 U:0.97V
read():1067 U:0.97V
```
The modified section:
```
import machine, esp32, time, os
from machine import Pin, Timer, ADC, DAC

led2 = Pin(2, Pin.OUT)    # create output pin on GPIO2
adcPin = 33
dacPin = 25
adc = ADC(Pin(adcPin))        # create an ADC object acting on a pin
adc.atten(ADC.ATTN_11DB) #max adc 2.6V
dac = DAC(Pin(dacPin))
Udac_max = 3.17 #max DAC voltage mesured
N = int(1.0/Udac_max*255)
dac.write(N)
vref = 1.039
vref11db = vref * (10**(11/20))


def timer(led):
    if led.value():
        led.value(0)# set pin to "off" (low) level
    else:
        led.value(1) # set pin to "on" (high) level
    
    print("read_uv():{:.2f}V".format(adc.read_uv()/1e6))
    N = adc.read_u16()
    print("readu16():{} U:{:.2f}V".format(N,(vref11db/(2**16))*N))
    N = adc.read()
    print("read():{} U:{:.2f}V".format(N, (vref11db/(2**12))*N))
```

**Reply by dvarrel** [reply]
i tried with other adc pin, and it seems my pin36 has problem. With pin33 or 32 it's ok :+1: 
using read_uv() use the vref value of the chip ?
if not, is there a solution to read vref voltage in efuse, like pyb library ( read_core_vref())
thanks

**Reply by davefes** [reply]
Which ESP32 variant, I have a note that `Pins 36, 37, 38 and 39 not used`.  Which dev board brings out pin36?

**Reply by dvarrel** [reply]
[ttgo t-display](https://pdacontrolen.com/wp-content/uploads/2021/07/screen37.jpg)

**Reply by davefes** [reply]
OK, my comment is for the generic ESP32.  Maybe yours is ESP32S2 or ESP32C3.  They show that is apparently usable.  Which firmware image have you got on it?

**Reply by dvarrel** [reply]
Chip is ESP32-D0WDQ6 (revision 1)
MicroPython v1.19.1-962-g4376c969f-dirty on 2023-03-09; TTGO T-Display with ESP32

**Reply by davefes** [reply]
The other note I have is:
`'GPIO34-39 can only be set as input mode and do not have software pullup or pulldown functions.'`
Seems to suggest that seeing as they have brought it out that you should be able to use that pin.
I see that it is a modified image, so unless they give you the details of the modifications then we are missing useful information.

Seems someone else had an issue with pin36
https://forum.micropython.org/viewtopic.php?f=18&t=4487&p=25968&hilit=pin36#p25968

**Reply by dvarrel** [reply]
on my other card, analog pin are shown like this : [wemos d1 R32](https://cdn.shopify.com/s/files/1/1509/1638/files/2_Pinout_D1_R32_1024x1024.png?v=1608323483)
can i MUX GPIO with micropython ? because ADC2 cause problems with WIFI

**Reply by davefes** [reply]
That is my understanding about ADC2.

> can i MUX GPIO with micropython ?

Do you mean change between GPIO input and output?  I hope so!  I change from Pin to Pin hold and back again in some watering timers.

Or change from ADC to GPIO in or out?
I think you would have to deinit the ADC block 
https://forum.micropython.org/viewtopic.php?f=18&t=10243
Good luck!

**Reply by dvarrel** [reply]
i mean IO_MUX changing
[IO_MUX and GPIO Matrix (GPIO, IO_MUX)](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf#iomuxgpio)

**Reply by davefes** [reply]
Sorry, I haven't been down to that level on the ESP32.
